### DISCUSSION TITLE
Native IFC and IfcOpenShell

### DISCUSSION THREAD

**Discussion by Martin15135215** [IfcOpenShell/IfcOpenShell#4215] [Q&A]
This is from [here](https://github.com/brunopostle/ifcmerge):

> 
> A Native IFC application behaves in the following ways when editing a pre-existing IFC (STEP/SPF) file:
> 
> - IFC entities must be written in the same format as received, with the same numeric IDs as before.
> 
> - Attribute changes to entities must be written in-place.
> 
> - Numeric IDs of deleted entities must not be reused for new entities.
> 

So the first therm is fine with IfcOpenShell, when an ifc file is read and written back, there are identical, with the step ID.

The second there is also fine with IfcOpenShell, since it also changes attribute in-place.

My question is, how does IfcOpenShell handle the last term?

So if I have e ifc file with step ID's from 1 to 10. I added a new entity with IfcOpenShell, it is written to the step ID 11.

If I now delete the entity at step ID 8. Write my IFC File to disk and read it in a new python instance of the ifc file and add a new entity, will it be at step ID 8 or step ID 12?

If instead of deleting step ID 8, I delete step ID 11. Then write my IFC file to disk and in a new python instance I open the ifc file with IfcOpenShell, and add a new entity, how does Ifcopenshell know that step ID 11 should not be used?

**Comment by dirkolbrich** [comment]
As far as I can read form the code, IfcOS has a MaxId counter in the IfcFile class:

https://github.com/IfcOpenShell/IfcOpenShell/blob/ce26c6e19b9acca4e8a5abb70c2e88fd8e146210/src/ifcparse/IfcFile.h#L146

During parsing of an IFC file, the current ID is read:

https://github.com/IfcOpenShell/IfcOpenShell/blob/ce26c6e19b9acca4e8a5abb70c2e88fd8e146210/src/ifcparse/IfcParse.cpp#L1619

and with each new Id the MaxId counter is incremented: 

https://github.com/IfcOpenShell/IfcOpenShell/blob/ce26c6e19b9acca4e8a5abb70c2e88fd8e146210/src/ifcparse/IfcParse.cpp#L1692

When adding a new Entity to the IFC model, the ID for the entity is generated by incrementing MaxId (function `FreshId()`): 

https://github.com/IfcOpenShell/IfcOpenShell/blob/ce26c6e19b9acca4e8a5abb70c2e88fd8e146210/src/ifcparse/IfcParse.cpp#L1999-L2011

> If instead of deleting step ID 8, I delete step ID 11. Then write my IFC file to disk and in a new python instance I open the ifc file with IfcOpenShell, and add a new entity, how does Ifcopenshell know that step ID 11 should not be used?

But I think your last use case is not covered, as during parsing the MaxId counter is only set to the maximum existing Id in the document. IfcOS can not look into the future.

I think if it is just 1 to a couple Id’s that should not be a problem. It becomes difficult If I would delete half a build, which Id’s happen to be at the end of the file.

—

Aside from that, I think that the cited requirements to be a NativeIFC application are not applicable, and even not correct. 

The requirements are specified in a white paper (a thesis) on how to use GIT to versioning (track changes, merging, diffing etc.) IFC files across multiple collaborators, just like currently used in writing code.

Instead the NativeIFC definition (IMHO) only means to work directly with an IFC file or data format while authoring building information and not just to use it as an intermediary exchange format between proprietary applications and their closed file formats like Revit.

Working with NativeIFC can be practised without using a versioning system. Just like I could send you a bunch HTML/CSS/JS files via email or physically on my good old floppy disc. I edited these files in a text editor like Vscodium, and you can continue the work using VIM in the terminal or whatever application you choose.

Another point is, that these requirements of not changing the numeric Id’s and not reusing unused ID would directly violate the definition of the STEP file format (an ISO standard), which IFC files are written in. Id’s, or in the context of STEP called instance names are only locally valid and can change on export or import by another application.

**Yet another point is: is GIT really suitable for IFC?** Git is primarily used (even intended?) for code files in plain text format, where the order of the lines matter. Also these files are to be interpreted by computers but also to a great part written and read by humans.

For raw IFC files this is not the case. Even though the file can be read by humans as it is written in plain text, its primarily use case is to be written and read by computers to allow interchangeability of data between different authoring applications. Directly viewing this data from an IFC file is not the main intend, as most of this data must be visually modified (3D model, plan, schedule etc.) to be consumed by humans.

You propably would never read a plain text DIFF view of the raw IFC file data to inspect what changed in the model, but use a prepared 3D view with color codes for existing/new/changed elements or an tabular data schedule.

For computers to track versions of IFC files it does not matter if the Step file ID is in the right order or reused from a previous state. IFC elements are identified by a GUID identifier, which should be enough to track changes (is the old element with the same GUID different than the current element? If yes, what changed and prepare it visually to be understood by a human).

I have a lot of respect for the work of Bruno Postle by creating that white paper and the ifcmerge application. It’s the best we currently have and I don’t have anything comparable. Yet I think the use of GIT in its current form for IFC data is flawed.

**Comment by Moult** [comment]
I think the answers you're looking for are here: https://github.com/IfcOpenShell/IfcOpenShell/issues/2240

TL;DR your concerns about deleted ID reuse is correct and @dirkolbrich 's concerns about SPF's specificity or Git's specificity are also already acknowledged in the above issue.

**Comment by aothms** [comment]
> But I think your last use case is not covered, as during parsing the MaxId counter is only set to the maximum existing Id in the document. IfcOS can not look into the future.
>
> I think if it is just 1 to a couple Id’s that should not be a problem. It becomes difficult If I would delete half a build, which Id’s happen to be at the end of the file.

Came here to say the same thing. IfcOpenShell indeed assigns new ids based on the highest id currently in the file. This was just lazyness, but turned out to be a feature. The only issue in this regard with ID reuse across successive write+read operations is indeed that if you delete the highest ID from the model, write to disk and read, then in that case specifically we *will recycle* that numeric id.

One possibly hacky way around this is to before doing a `file.write()`, do a `file.createIfcPerson(Id='<end of file>')` or something equally ugly. To make sure basically  that the maxid is propagated even when the highest id in the population being written to disk is lower. You could also persist the value of maxid into some comment I guess.

**Comment by Martin15135215** [comment]
Thanks, @dirkolbrich @Moult and @aothms for the inputs!

What a Native IFC is supposed to be is therefore clear:

> We propose a new paradigm, creating and editing IFC data in-place without import/export translation to proprietary models: Native IFC. Source: [Native IFC Whitepaper](https://github.com/brunopostle/ifcmerge/blob/14807c48506c11f0ccb22b8e625a254ed436202b/docs/whitepaper.rst?plain=1#L52)

> Instead the NativeIFC definition (IMHO) only means to work directly with an IFC file or data format while authoring building information and not just to use it as an intermediary exchange format between proprietary applications and their closed file formats like Revit. Source: [here](https://github.com/IfcOpenShell/IfcOpenShell/discussions/4215#discussioncomment-8133507)

The advantages would be round-tripping without data loss that would otherwise happen in other OpenBIM workflow.

The other advantages of using git with ifc would fail in the last described edge case and would need some work-around(?). Is there an effort in the Next IFC Gen (IFC5) to make it different, to make IFC more friendly for direct use and not only functional as an exchange format? Is it currently not possible to model parametrically (?) in IFC? I mean, like save the parameter?

**Reply by aothms** [reply]
> Is there an effort in the Next IFC Gen (IFC5) to make it different, to make IFC more friendly for direct use and not only functional as an exchange format

Definetely. It seems to gravitating towards an Entity-component structure. Where components can be individually versioned and collaborative assembled "into" an entity.

**Reply by brunopostle** [reply]
> One possibly hacky way around this is to before doing a file.write(), do a file.createIfcPerson(Id='<end of file>') or something equally ugly. To make sure basically that the maxid is propagated even when the highest id in the population being written to disk is lower. You could also persist the value of maxid into some comment I guess.

ifcmerge will refuse to do the merge if these trailing IDs have been recycled, so at least it shouldn't result in corrupt data.

I don't think it will be necessary to do any hacky stuff to propagate the maxid to the next session - ifcmerge could detect this situation and rewrite the IDs to be what they should have been with no recycling. I'm inclined to only worry about this when it actually causes problems, but if somebody wants to add this feature then that would be cool too.

Of course this way of working is not relevant in a world where applications are doing import/export, but all the alternative ideas for open collaboration, such as a centralised database-server, require that everyone strictly uses the same schema and entity creation/modification/deletion is tracked - ie. some form of Native IFC like thing.

In some ways IFC is a better fit with git than source-code is. The step-IDs really help - imagine if programming languages still had line numbers like BASIC, there would be no need for fuzzy matching and all the offset searching that happens when lines are inserted and deleted - the numbers add certainty.
